### ISSUE TITLE
crazy proposal: parse diffs to figure out when re-review is required

### ISSUE THREAD

**Issue by jeffkaufman** [jeffkaufman/nomic#95]
Right now we can avoid re-reviews in the situation where git can automatically handle the merge conflicts, which we handle by simply not merging master into the PR.  But sometimes we need to merge master in or else the PR won't pass Travis: consider https://github.com/jeffkaufman/nomic/pull/93 where I needed the PR to run on python 3.6 or https://github.com/jeffkaufman/nomic/pull/67 where the branch in the PR was looking for approvals from people who were no longer players.  Both needed re-review after merging in master, even though the PR diff didn't change at all.

If this gets annoying we should be able to make something that allows approvals from previous commits to still stand under very specific restrictions.  Something like, if the diff "approved commit vs master at the time of approval" is identical to "current commit vs master now".

**Comment by tnelling** [comment]
I can imagine situations where we wouldn't want this behavior. Not so much because of what merging master does to the PR, but because (perhaps) something having been merged into master changes the higher-level context in which the PR is being evaluated. Take the example of the discussion about whether named bonuses should have the values in the file contents or file names. The PRs did in fact step on each other, but it wouldn't be hard to imagine a similar situation where they didn't, but nonetheless, once one of those two PRs went in, the other should not.

That all having been said, maintaining current approvals wouldn't necessarily cause such a PR to go in anyway, and in the vast majority of cases what you're suggesting makes sense.

**Comment by jeffkaufman** [comment]
@tnelling sure, but we have that problem now.  For example, #47 and #68/#69 don't have any merge conflicts and github would have allowed #68/#69 and then #47 to merge.  But unless #47 were synced after #68/#69 were merged (which I intentionally did) then #47 would have added the two players back into the game.

**Comment by tnelling** [comment]
True. You could even argue we should go the complete other direction - all approvals are wiped after anything is merged into master. I think we may have already discussed that a bit, actually, as a hypothetical opening for sneaky behavior. I probably wouldn't even mind having to re-approve, but I'd be concerned about what happens when other players don't do so.

**Comment by jeffkaufman** [comment]
That's how we had things when we very first started playing, but we turned it off because it was too annoying: https://github.com/jeffkaufman/nomic/issues/18

I don't think it's worth it.

**Comment by jeffkaufman** [comment]
Here's how this could work: an approval at commit N is invalidated by pushing a new commit M unless the PR diff is *exactly the same* at commits N and M.  In other words, merging from master would reset approvals if the files in question had been touched (added, removed, or edited) on master, because that would affect the left side of the diff.

Automated PR checks would still need to pass, of course.

**Comment by jeffkaufman** [comment]
(this is significantly less crazy than I had been imagining)
